### DISCUSSION TITLE
deepgram.errors.DeepgramApiError: <urlopen error EOF occurred in violation of protocol (_ssl.c:2427)>

### DISCUSSION THREAD

**Discussion by redshiva** [deepgram/community#436] [General help]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

Requesting transcript... for  /home/dave/Downloads/harvard.wav
Traceback (most recent call last):
  File "/usr/lib/python3.11/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.11/http/client.py", line 1286, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1332, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1281, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.11/http/client.py", line 1080, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.11/http/client.py", line 1002, in send
    self.sock.sendall(data)
  File "/usr/lib/python3.11/ssl.py", line 1274, in sendall
    v = self.send(byte_view[count:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/ssl.py", line 1243, in send
    return self._sslobj.write(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2427)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dave/Dropbox/Code/LLM/venv/lib/python3.11/site-packages/deepgram/_utils.py", line 203, in _sync_request
    return attempt()
           ^^^^^^^^^
  File "/home/dave/Dropbox/Code/LLM/venv/lib/python3.11/site-packages/deepgram/_utils.py", line 177, in attempt
    with urllib.request.urlopen(req, timeout=timeout) as resp:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 519, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 1391, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/request.py", line 1351, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error EOF occurred in violation of protocol (_ssl.c:2427)>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/dave/Dropbox/Code/LLM/Media-Ingest/test.py", line 28, in <module>
    main()
  File "/home/dave/Dropbox/Code/LLM/Media-Ingest/test.py", line 23, in main
    response = dg_client.transcription.sync_prerecorded(source, options)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dave/Dropbox/Code/LLM/venv/lib/python3.11/site-packages/deepgram/transcription.py", line 378, in sync_prerecorded
    return SyncPrerecordedTranscription(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dave/Dropbox/Code/LLM/venv/lib/python3.11/site-packages/deepgram/transcription.py", line 121, in __call__
    return _sync_request(
           ^^^^^^^^^^^^^^
  File "/home/dave/Dropbox/Code/LLM/venv/lib/python3.11/site-packages/deepgram/_utils.py", line 209, in _sync_request
    raise DeepgramApiError(exc, http_library_error=exc) from exc
deepgram.errors.DeepgramApiError: <urlopen error EOF occurred in violation of protocol (_ssl.c:2427)>

Process finished with exit code 1

### If possible, please attach your code or paste it into the text box.

from deepgram import Deepgram
import json
DEEPGRAM_API_KEY = ''
PATH_TO_FILE = '/home/dave/Downloads/harvard.wav'
MIMETYPE = 'audio/wav'

def main():
    # Initializes the Deepgram SDK
    dg_client = Deepgram(DEEPGRAM_API_KEY)

    with open(PATH_TO_FILE, 'rb') as audio:
        source = {'buffer': audio, 'mimetype': MIMETYPE}
        options = {"smart_format": False, "model": "nova", "language": "en-GB"}
        print('Requesting transcript... for ', PATH_TO_FILE)
        response = dg_client.transcription.sync_prerecorded(source, options)
        print(json.dumps(response, indent=4))

main()

### If possible, please attach an example audio file to reproduce the issue.

[harvard.zip](https://github.com/deepgram/community/files/13327076/harvard.zip)

**Comment by jkroll-deepgram** [comment]
Hi @enkiman, your code and audio file worked as-is for me. I'm using Python 3.11.5 and Deepgram Python SDK 2.10.0.

So I can attempt to reproduce, can you share which versions you're using? Based on the error, it does look like you're using some version of Python 3.11.

To check your Deepgram version in the Python interpreter, you can use:
```
>>> import deepgram
>>> deepgram._version.__version__
```

**Reply by redshiva** [reply]
I'm using python 3.11.6
deepgram 2.12.0

**Reply by jkroll-deepgram** [reply]
I bumped my Deepgram version up to 2.12.0 and it still ran smoothly for me. I suspect this is an issue with your environment, given the error `ssl.SSLEOFError: EOF occurred in violation of protocol`. Are you using any kind of VPN, virtual environment, unusual networking? If not, I recommend searching online for that error, there are a few packages and configurations that could be causing it.

For reference, here's a couple other issues with similar reported errors; they may not be your particular error, but could raise some considerations to help you debug.
https://github.com/orgs/deepgram/discussions/374
https://github.com/orgs/deepgram/discussions/132

**Comment by redshiva** [comment]
Thank you for the prompt reply.  The two articles referenced, do not quite describe my situation.  The fix for the first one was a corrupted audio file.  You indicated that you were able to transcribe my file without issue.  The second one was about a mismatch of parameters when calling the API, again, not an issue here because you were able to run my code in your environment without issue.

I do not have any exotic network going on here, just a workstation behind a firewall; no VPN.   I am using a venv managed by pycharm.

Here is a listing of the packages and versions in the venv:
aiohttp	3.8.6
aiosignal	1.3.1
appdirs	1.4.4
apt-xapian-index	0.49
async-timeout	4.0.3
attrs	23.1.0
beautifulsoup4	4.12.2
beniget	0.4.1
blinker	1.7.0
bottle	0.12.25
Brotli	1.1.0
certifi	2023.7.22
cffi	1.16.0
chardet	5.2.0
charset-normalizer	3.3.2
click	8.1.7
colorama	0.4.6
command-not-found	0.3
contourpy	1.2.0
cryptography	41.0.5
cupshelpers	1
cycler	0.12.1
dbus-python	1.3.2
decorator	5.1.1
deepgram-sdk	2.12.0
defusedxml	0.8.0rc2
distro	1.8.0
distro-info	1.5
docker	6.1.3
fonttools	4.44.1
frozenlist	1.4.0
fs	2.4.16
fuse-python	1.0.5
gast	0.5.4
Glances	3.4.0.3
gpg	1.18.0
html5lib	1.1
httplib2	0.22.0
idna	3.4
importlib-metadata	6.8.0
influxdb	5.3.1
jaraco.classes	3.3.0
jeepney	0.8.0
keyring	24.3.0
kiwisolver	1.4.5
language-selector	0.1
launchpadlib	1.11.0
lazr.restfulclient	0.14.5
lazr.uri	1.0.6
lxml	4.9.3
lz4	4.3.2
markdown-it-py	3.0.0
matplotlib	3.8.1
mdurl	0.1.2
more-itertools	10.1.0
mpmath	1.3.0
msgpack	1.0.7
multidict	6.0.4
mutagen	1.47.0
ndg-httpsclient	0.5.1
netifaces	0.11.0
numpy	1.26.2
oauthlib	3.2.2
olefile	0.47.dev4
packaging	23.2
pexpect	4.8.0
Pillow	10.1.0
pip	23.3.1
ply	3.11
psutil	5.9.6
psycopg2-binary	2.9.9
ptyprocess	0.7.0
pyasn1	0.5.0
pycairo	1.24.0
pycparser	2.21
pycryptodomex	3.19.0
pycups	2.0.1
Pygments	2.16.1
PyGObject	3.46.0
PyJWT	2.8.0
pylibacl	0.7.0
pyOpenSSL	23.3.0
pyparsing	3.1.1
PyQt5	5.15.10
PyQt5-Qt5	5.15.2
PyQt5-sip	12.13.0
pysmi	0.3.4
pysnmp	4.4.12
pystache	0.6.5
python-apt	2.6.0+ubuntu1
python-dateutil	2.8.2
python-debian	0.1.49+ubuntu2
python-json-logger	2.0.7
pythran	0.14.0
pytz	2023.3.post1
pyxattr	0.8.1
PyYAML	6.0.1
requests	2.31.0
rich	13.6.0
scipy	1.11.3
SecretStorage	3.3.3
setuptools	68.2.2
six	1.16.0
soupsieve	2.5
sympy	1.12
systemd-python	235
tornado	6.3.3
ubuntu-advantage-tools	8001
ubuntu-drivers-common	0.0.0
ufoLib2	0.16.0
ufw	0.36.2
ujson	5.8.0
unattended-upgrades	0.1
urllib3	2.1.0
usb-creator	0.3.16
wadllib	1.3.6
webencodings	0.5.1
websocket-client	1.6.4
websockets	12
wheel	0.41.3
xkit	0.0.0
yarl	1.9.2
yt-dlp	2023.11.14
zipp	3.17.0

Can you provide me with a list of all of the python packages and versions you are using in your 3.11.6/2.12.0 environment?

**Reply by jkroll-deepgram** [reply]
Here's the output of `pip3 freeze > requirements.txt` in my environment:
```
aiohttp==3.8.4
aiosignal==1.3.1
alembic==1.11.1
anyio==3.7.0
appnope==0.1.3
argilla==1.7.0
asttokens==2.2.1
async-timeout==4.0.2
asyncpg==0.27.0
attrs==22.2.0
Automat==22.10.0
backcall==0.2.0
backoff==2.2.1
beautifulsoup4==4.12.2
bs4==0.0.1
certifi==2023.5.7
cffi==1.15.1
chardet==4.0.0
charset-normalizer==3.1.0
click==8.1.3
comm==0.1.3
commonmark==0.9.1
constantly==15.1.0
coverage==7.2.7
cryptography==40.0.1
cssselect==1.2.0
Cython==3.0.0
databases==0.6.2
dataclasses-json==0.5.7
debugpy==1.6.7
decorator==5.1.1
deepgram-sdk==2.12.0
Deprecated==1.2.13
dnspython==2.3.0
email-validator==2.0.0.post2
et-xmlfile==1.1.0
exceptiongroup==1.1.1
executing==1.2.0
fastapi==0.95.2
ffmpeg-python==0.2.0
filelock==3.11.0
flatten-json==0.1.13
frozenlist==1.3.3
fsspec==2023.6.0
future==0.18.3
greenlet==2.0.2
grpcio==1.54.2
grpcio-tools==1.54.2
h11==0.14.0
h2==4.1.0
hpack==4.0.0
httpcore==0.17.2
httptools==0.5.0
httpx==0.24.1
huggingface-hub==0.15.1
hyperframe==6.0.1
hyperlink==21.0.0
idna==3.4
incremental==22.10.0
iniconfig==2.0.0
ipykernel==6.23.2
ipython==8.14.0
isodate==0.6.1
itemadapter==0.8.0
itemloaders==1.0.6
itsdangerous==2.1.2
jedi==0.18.2
Jinja2==3.1.2
jmespath==1.0.1
joblib==1.2.0
jupyter_client==8.2.0
jupyter_core==5.3.1
langchain==0.0.173
lxml==4.9.2
Mako==1.2.4
Markdown==3.4.3
markdownify==0.11.6
MarkupSafe==2.1.2
marshmallow==3.19.0
marshmallow-enum==1.5.1
matplotlib-inline==0.1.6
monotonic==1.6
msg-parser==1.2.0
multidict==6.0.4
mypy-extensions==1.0.0
nest-asyncio==1.5.6
nltk==3.8.1
numexpr==2.8.4
numpy==1.24.3
olefile==0.46
openai==0.27.6
openapi-schema-pydantic==1.2.4
openpyxl==3.1.2
orjson==3.8.12
ormar==0.12.1
packaging==23.1
pandas==1.5.3
parsel==1.7.0
parso==0.8.3
pdf2image==1.16.3
pdfminer.six==20221105
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.5.0
platformdirs==3.7.0
pluggy==1.2.0
portalocker==2.7.0
prompt-toolkit==3.0.38
Protego==0.2.1
protobuf==4.23.3
psutil==5.9.5
psycopg2-binary==2.9.6
ptyprocess==0.7.0
pure-eval==0.2.2
pyasn1==0.4.8
pyasn1-modules==0.2.8
PyAudio==0.2.13
pybind11==2.11.1
pycairo==1.24.0
pycparser==2.21
pydantic==1.10.9
PyDispatcher==2.0.7
pydub==0.25.1
Pygments==2.15.1
PyGObject==3.44.1
PyJWT==2.8.0
pyOpenSSL==23.1.1
pypandoc==1.11
pytest==7.4.0
pytest-cov==4.1.0
pytest-dotenv==0.5.2
pytest-env==0.8.1
pytest-mock==3.10.0
python-dateutil==2.8.2
python-docx==0.8.11
python-dotenv==1.0.0
python-graphql-client==0.4.3
python-magic==0.4.27
python-multipart==0.0.6
python-pptx==0.6.21
pytz==2023.3
PyYAML @ file:///private/tmp/pyyaml-20230717-6270-1iiwi1j/PyYAML-6.0.1
pyzmq==25.1.0
qdrant-client==1.2.0
queuelib==1.6.2
regex==2023.6.3
requests==2.31.0
requests-file==1.5.1
requests-toolbelt==1.0.0
rfc3986==1.5.0
rich==13.0.1
safetensors==0.3.1
schedule==1.2.0
Scrapy==2.8.0
service-identity==21.1.0
simple-salesforce==1.12.4
six==1.16.0
slack-bolt==1.18.0
slack-sdk==3.21.3
sniffio==1.3.0
soupsieve==2.4.1
SpeechRecognition==3.10.0
SQLAlchemy==1.4.41
stack-data==0.6.2
starlette==0.27.0
tabulate==0.9.0
tenacity==8.2.2
tiktoken==0.4.0
tldextract==3.4.0
tokenizers==0.13.3
tomli==2.0.1
tornado==6.3.2
tqdm==4.65.0
traitlets==5.9.0
transformers==4.30.1
Twisted==22.10.0
typer==0.9.0
typing-inspect==0.8.0
typing_extensions==4.5.0
ujson==5.7.0
unstructured==0.6.7
urllib3==1.26.16
uvicorn==0.22.0
uvloop==0.17.0
w3lib==2.1.1
watchfiles==0.19.0
wcwidth==0.2.6
websockets==11.0.3
wrapt==1.14.1
XlsxWriter==3.1.0
yarl==1.9.2
zeep==4.2.1
zope.interface==6.0
```

**Comment by davidvonthenen** [comment]
What OS are you on?

**Comment by redshiva** [comment]
ubuntu 23.10



________________________________

**Comment by redshiva** [comment]
I solved this issue.  I updated some libraries and changed the audio file and all is well now!

Thank you all!
